src/features/APIWithHooks.ts

```diff
@@ -0,0 +1,61 @@
+import {IAPI} from '../APICore';
+
+export interface IAPIHooks {
+    beforeAnyRequest?: (url: string, args: RequestInit) => void;
+    afterAnySuccess?: <T>(url: string, args: RequestInit, response: T) => T;
+    afterAnyException?: (url: string, args: RequestInit, exception: Error) => boolean;
+}
+
+export class APIWithHooks<TAPI extends IAPI = IAPI> implements IAPI {
```

Comment:
How would that class integrates with the current client? I don't see it used anywhere yet.

Reply:
Would it be simpler to directly put `beforeAnyRequest`, `afterAnySuccess` and `afterAnyException` directly inside the API class?

Reply:
The code would be simpler, but testing `API` would require exponentially more tests to ensure all use cases are covered. 

I like the way it is built now just because the two concerns (basic API vs API with hooks) are clearly defined.